File: backend/node/src/aws/Translate.mjs
import {
  ComprehendClient,
  DetectDominantLanguageCommand,
} from "@aws-sdk/client-comprehend";
import {
  TranslateClient,
  TranslateTextCommand,
} from "@aws-sdk/client-translate";
import config from "../config.mjs";

// Configura el cliente de AWS Comprehend
const comprehendClient = new ComprehendClient({
  region: config.region, // Reemplaza con tu región
  credentials: {
    accessKeyId: config.accessKeyId_comprehend, // O tu ACCESS_KEY_ID directamente
    secretAccessKey: config.secretAccessKey_comprehend, // O tu SECRET_ACCESS_KEY directamente
  },
});

// Configura el cliente de AWS Translate
const translateClient = new TranslateClient({
  region: config.region, // Reemplaza con tu región
  credentials: {
    accessKeyId: config.accessKeyId_translate, // O tu ACCESS_KEY_ID directamente
    secretAccessKey: config.secretAccessKey_translate, // O tu SECRET_ACCESS_KEY directamente
  },
});

const TranslateText = async (text, targetLang) => {
  try {
    // Detecta el idioma del texto
    const detectCommand = new DetectDominantLanguageCommand({ Text: text });
    const detectResponse = await comprehendClient.send(detectCommand);
    const sourceLang = detectResponse.Languages[0].LanguageCode;
    console.log(`Detected Language: ${sourceLang}`);

    // Traduce el texto
    const translateCommand = new TranslateTextCommand({
      Text: text,
      SourceLanguageCode: sourceLang,
      TargetLanguageCode: targetLang,
    });
    const translateResponse = await translateClient.send(translateCommand);
    //console.log(`Translated Text: ${translateResponse.TranslatedText}`);
    return translateResponse.TranslatedText;
  } catch (error) {
    console.error(error);
    return null;
  }
};
//detectAndTranslateText("Hola, ¿cómo estás?", "en");
export default TranslateText;

File: backend/python/src/aws/Translate.py
import boto3
from botocore.exceptions import BotoCoreError, ClientError
from config import config

# Crear cliente de AWS Comprehend
comprehend_client = boto3.client(
    'comprehend',
    region_name=config['region_ha'],
    aws_access_key_id=config['accessKeyId_comprehend'],
    aws_secret_access_key=config['secretAccessKey_comprehend']
)

# Crear cliente de AWS Translate
translate_client = boto3.client(
    'translate',
    region_name=config['region_ha'],
    aws_access_key_id=config['accessKeyId_translate'],
    aws_secret_access_key=config['secretAccessKey_translate']
)

def translate_text(text, target_lang):
    try:
        # Detectar el idioma del texto
        detect_response = comprehend_client.detect_dominant_language(Text=text)
        source_lang = detect_response['Languages'][0]['LanguageCode']
        print(f"Detected Language: {source_lang}")

        # Traducir el texto
        translate_response = translate_client.translate_text(
            Text=text,
            SourceLanguageCode=source_lang,
            TargetLanguageCode=target_lang
        )

        return translate_response['TranslatedText']
    except (BotoCoreError, ClientError) as error:
        print(f"Error: {error}")
        return None
